make/photon/prepare/utils/migration.py

```diff
@@ -15,29 +32,35 @@ def read_conf(path):
             exit(-1)
     return d
 
-def _to_module_path(ver):
-    return "migrations.version_{}".format(ver.replace(".","_"))
-
 def search(input_ver: str, target_ver: str) -> deque :
     """
-    Search accept a input version and the target version.
-    Returns the module of migrations in the upgrade path
+    Find the migration path by BFS
+    Args:
+        input_ver(str): The version migration start from
+        target_ver(str): The target version migrated to
+    Returns:
+        deque: the module of migrations in the upgrade path
     """
-    upgrade_path, visited = deque(), set()
-    while True:
-        module_path = _to_module_path(target_ver)
-        visited.add(target_ver)  # mark current version for loop finding
-        if os.path.isdir(os.path.join(MIGRATION_BASE_DIR, 'version_{}'.format(target_ver.replace(".","_")))):
-            module = importlib.import_module(module_path)
-            if module.revision == input_ver:    # migration path found
-                break
-            elif module.down_revision is None:  # migration path not found
-                raise Exception('no migration path found')
-            else:
-                upgrade_path.appendleft(module)
-                target_ver = module.down_revision
-                if target_ver in visited: # version visited before, loop found
-                    raise Exception('find a loop caused by {} on migration path'.format(target_ver))
-        else:
-            raise Exception('{} not dir'.format(os.path.join(MIGRATION_BASE_DIR, 'versions', target_ver.replace(".","_"))))
-    return upgrade_path
+    upgrade_path = deque()
+    parents, visited, q = dict(), set(), deque() # must ensure all version in parents are valid
+    q.append(target_ver)
+    while q: # BFS to find a valid path
+        version = MigrationVersion(q.popleft())
+        if version.name in visited:
+            continue
+        visited.add(version.name)
+        if version.name == input_ver: 
+            upgrade_path.append(version)
+            break # break loop cause migration path found
+        for v in version.down_versions:
+            parents[v] = version.name
+            q.append(v)
+
+    if not upgrade_path:
+        raise MigratioNotFound('no migration path found to target version')
+
+    current_version = upgrade_path.popleft()
```

Comment:
I find it a little confusing you append the input_ver into upgrade_path and pop it and never put it back again.

If you use a flag like `found` in line 59 here you only need to assign `input_ver` to `current_version`, right?

Reply:
Because in this scenario the last element in `upgrade_path` can be used as a flag. Or you need an extra flag variable and initialize a MigrateVersion instance in another place.